### PR TITLE
wip irrlicht particles 2

### PR DESCRIPTION
**Pull request by obneq**
next attempt here, using custome scene node: 
https://github.com/obneq/minetest/tree/scenenodeparticles

sorry for closing and reopening instead of updating PR. this is attempt 2 at a naive replacement of mt particles with irrlicht ones. 
it does not yet look like current mt particles and is not meant for straight merge. Instead, developers please review and give your feedback! Please note that this is very much WIP. Ignore style and minor issues if you can and instead let me know what could be done better differently in a larger sense.

### CODE REVIEW

**Comment by C1ffisme** [comment]
What advantages does it pose? Does it make them faster/better?

**Comment by kwolekr** [comment]
Both faster and better.  This is using the builtin Irrlicht particle functionality which we've been waiting for :)

**Comment by nerzhul** [comment]
The original pr has collision bugs and generate some lag on Intel GPU

**Comment by nerzhul** [comment]
Then i follow it with attention :)

**Comment by Zeno-** [comment]
I'm not sure how I missed this PR. Watching.

File: src/particles.cpp [hunk 1]
```diff
@@ -339,235 +271,174 @@ ParticleManager::~ParticleManager()
...
+				gamedef->tsrc()->getTextureForMesh(*(event->add_particlespawner.texture));
+
+		scene::IParticleSystemSceneNode * ps = m_smgr->addParticleSystemSceneNode();
+
+		ps->setID((s32) event->add_particlespawner.id);
+
+		float pps = event->add_particlespawner.amount;
+		float time = event->add_particlespawner.spawntime;
+
+		if (time != 0)
```
Comment (Zeno-) [inline src/particles.cpp]: Please don't compare floating point numbers to an absolute number

File: src/particles.cpp [hunk 2]
```diff
@@ -44,294 +49,221 @@ v3f random_v3f(v3f min, v3f max)
...
+        }
+
+        void affect(u32 now, irr::scene::SParticle* particlearray, u32 count)
+        {
+                if (count<=0) {
+                        SceneManager->addToDeletionQueue(ParticleSystem);
+                }
+        }
+        irr::scene::E_PARTICLE_AFFECTOR_TYPE getType() const {
+                return (irr::scene::E_PARTICLE_AFFECTOR_TYPE) 667;
```
Comment (Zeno-) [inline src/particles.cpp]: There is no existing enum for whatever 667 is?

File: src/particles.cpp [hunk 3]
```diff
@@ -44,294 +49,221 @@ v3f random_v3f(v3f min, v3f max)
...
-	float expirationtime,
-	float size,
-	bool collisiondetection,
-	bool vertical,
-	video::ITexture *texture,
-	v2f texpos,
-	v2f texsize
-):
-	scene::ISceneNode(smgr->getRootSceneNode(), smgr)
+class fixNumEmitter :  public irr::scene::IParticleEmitter {
```
Comment (Zeno-) [inline src/particles.cpp]: I think all class names should begin with a capital letter

**Comment by kilbith** [comment]
So I tested that PR with your latest commit (4470c27) and here's my report : 

##### 1. Compiler's warnings

```
/home/jp/git/irrlicht-particles/minetest/src/particles.cpp: In member function ‘virtual const vector3df& FixNumEmitter::getDirection() const’:
/home/jp/git/irrlicht-particles/minetest/src/particles.cpp:108:87: warning: returning reference to temporary [-Wreturn-local-addr]
         virtual const core::vector3df& getDirection() const { return v3f(0.0, 0.0, 0.0); }
                                                                                       ^
/home/jp/git/irrlicht-particles/minetest/src/particles.cpp: In member function ‘virtual const irr::video::SColor& FixNumEmitter::getMinStartColor() const’:
/home/jp/git/irrlicht-particles/minetest/src/particles.cpp:111:87: warning: returning reference to temporary [-Wreturn-local-addr]
         virtual const video::SColor& getMinStartColor() const { return video::SColor(1); }
                                                                                       ^
/home/jp/git/irrlicht-particles/minetest/src/particles.cpp: In member function ‘virtual const irr::video::SColor& FixNumEmitter::getMaxStartColor() const’:
/home/jp/git/irrlicht-particles/minetest/src/particles.cpp:112:87: warning: returning reference to temporary [-Wreturn-local-addr]
         virtual const video::SColor& getMaxStartColor() const { return video::SColor(1);; }
                                                                                       ^
/home/jp/git/irrlicht-particles/minetest/src/particles.cpp: In member function ‘virtual const dimension2df& FixNumEmitter::getMaxStartSize() const’:
/home/jp/git/irrlicht-particles/minetest/src/particles.cpp:113:107: warning: returning reference to temporary [-Wreturn-local-addr]
         virtual const core::dimension2df& getMaxStartSize() const { return core::dimension2d<f32>(0.2, 0.2); }
                                                                                                           ^
/home/jp/git/irrlicht-particles/minetest/src/particles.cpp: In member function ‘virtual const dimension2df& FixNumEmitter::getMinStartSize() const’:
/home/jp/git/irrlicht-particles/minetest/src/particles.cpp:114:107: warning: returning reference to temporary [-Wreturn-local-addr]
         virtual const core::dimension2df& getMinStartSize() const { return core::dimension2d<f32>(0.2, 0.2); }
                                                                                                           ^
/home/jp/git/irrlicht-particles/minetest/src/particles.cpp: In member function ‘virtual const vector3df& FixNumEmitter::getCenter() const’:
/home/jp/git/irrlicht-particles/minetest/src/particles.cpp:115:84: warning: returning reference to temporary [-Wreturn-local-addr]
         virtual const core::vector3df& getCenter() const { return v3f(0.0, 0.0, 0.0); }
                                                                                    ^
```

##### 2. In-game comparison

Particles on current master (5a1975a4) - they have the good direction.
screenshot_20150511_150136

Same configuration with your Irrlicht particles branch - particles are spreading randomly everywhere surrounded by a moving white box - performances seems similar to current master.
screenshot_2015-05-11t15 00 30
screenshot_2015-05-11t15 15 56

**Comment by Zeno-** [comment]
Yes, well. Those warnings cannot be disregarded at all.

**Comment by paramat** [comment]
@obneq perhaps someone could adopt this work and continue it? Or are you planning to restart work? MT particles really need improvement.

**Comment by nerzhul** [comment]
Definitively need huge rework, possible close label

**Pull request closed without merging**